### PR TITLE
Voice reset code, JS deploy scripts added

### PR DESCRIPTION
**Pull request by n13**
## Voice reset
action voicereset added

## Tests
test/voice.test.js added

Contains comprehensive test for the voicereset action. All passing.

## JS build system
scripts/hypha.js added, and others

I was unable to use cmake since it has an error on M1 macs where it screws up the processor architecture. Only the test make is affected. But this meant I couldn't write tests. 

I believe the JS build system is better since it has easy commands to 
- build
- deploy
- test
- init test environment / contracts

It also builds ABI files and does not rely on pre-created abi files (although maybe the cmake based system does this too)

Certainly the cmake system has no way to deploy, that must be done by hand via cleos. 

### usage
./scripts/hypha.js run voice -> compiles and deploys

## Issues [TODO]
I had to change the imports and copying the document-graph files into our src/include directories

The git build script does not do that hence the cmake build fails

This is not good.

### CODE REVIEW

**Comment by n13** [comment]
Current issues

- Build is failing here

Potential fix: Change the cmake scripts to copy the files the same way the JS  copies the files into src and include (@IanMendozaJaimes )

- Use CMake instead of "compile" inside the JS script - That way it's compatible with normal cMake method of building and this build also won't fail (@IanMendozaJaimes )

- Use klevoya for unit tests (@IanMendozaJaimes )

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "document-graph"]
+	path = document-graph
+	url = git@github.com:hypha-dao/document-graph.git
```
Comment (josejulio) [inline .gitmodules:3]: Instead of using gitmodules, I'm using cmake to fetch the dependencies.

https://github.com/hypha-dao/voice-token/blob/master/src/CMakeLists-ExternalProjects.txt#L3-L7

We can remove the `gitsubmodules`.

File: Testing/Temporary/LastTest.log [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Start testing: Dec 20 21:02 CST
+----------------------------------------------------------
+End testing: Dec 20 21:02 CST
```
Comment (josejulio) [inline Testing/Temporary/LastTest.log]: Is this file needed? Else we can add it to the `.gitignore`

File: Testing/Temporary/CTestCostData.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+---
```
Comment (josejulio) [inline Testing/Temporary/CTestCostData.txt]: Is this file needed? Else we can add it to the `.gitignore`

File: include/decay.hpp [hunk 4]
```diff
@@ -15,7 +15,7 @@ namespace hypha {
         uint64_t newPeriod;
     };
 
-    const DecayResult decay(
+    const DecayResult Decay(
```
Comment (josejulio) [inline include/decay.hpp]: Are we following any naming standard / strategy ?

File: scripts/compile.js [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+const command = ({ contract, source, include, dir }) => {
+  const volume = dir
+  let cmd = ""
+  let inc = include == "" ? "./include" : include
+  
+  if (process.env.COMPILER === 'local') {
+    cmd = "eosio-cpp -abigen -I "+ inc +" -contract " + contract + " -o ./artifacts/"+contract+".wasm "+source;
+  } else {
+    cmd = `docker run --rm --volume ${volume}:${volume} -w ${volume} eosio/key-value-example:v1.0.0 /bin/bash -c "cd build && cmake .. && make"`
```
Comment (josejulio) [inline scripts/compile.js:23]: I'm using other image for building - I don't mind using either one, but we should probably stick to one: https://github.com/hypha-dao/voice-token/blob/master/.github/workflows/build.yml#L9

Also, we could probably "glue" this part with cmake, There could be a way (not sure) to tell cmake to build and execute this test. It would be cleaner IMO

edit: Ah, it doesn't look like it's a test. Anyhow we should use only one docker builder

File: src/voice.cpp [hunk 6]
```diff
@@ -186,6 +186,70 @@ namespace hypha {
...
+        const auto from = from_acnts.find( hvoice_symbol.code().raw());
+
+        if (from == from_acnts.end()) {
+            // No balance exists yet, nothing to do
+            return;
+        }
+        
+        if (from->balance > cutoff_value) {
+            eosio::asset old_balance = from->balance; //10
+            eosio::asset new_balance(5000 * 100, hvoice_symbol); // 1
```
Comment (josejulio) [inline src/voice.cpp:210]: Is this OK? 5000.00 HVOICE

**Comment by josejulio** [comment]
> I was unable to use cmake since it has an error on M1 macs where it screws up the processor architecture. Only the test make is affected. But this meant I couldn't write tests.

Could you try with the docker image I use in the workflow? 

> I believe the JS build system is better since it has easy commands to

Maybe we could use the JS only for deploying or as a "front facing" - Anyway the js relies on cmake/docker to run.

> It also builds ABI files and does not rely on pre-created abi files (although maybe the cmake based system does this too)

It does build the ABI files

> I had to change the imports and copying the document-graph files into our src/include directories
> The git build script does not do that hence the cmake build fails
> This is not good.

`cmake` does that (pulling the document graph)

File: Testing/Temporary/CTestCostData.txt (hunk 3 above)
Comment (n13) [reply Testing/Temporary/CTestCostData.txt]: I didn't notice these - IDK what created them ;)

File: include/decay.hpp (hunk 4 above)
Comment (n13) [reply include/decay.hpp]: I did not think I changed that...

File: src/voice.cpp (hunk 6 above)
Comment (n13) [reply src/voice.cpp:210]: YES

File: include/decay.hpp (hunk 4 above)
Comment (n13) [reply include/decay.hpp]: @IanMendozaJaimes

File: scripts/compile.js (hunk 5 above)
Comment (n13) [reply scripts/compile.js:23]: I don't mind changing this to anything else

@josejulio uses docker://josejulio/hypha-builder:release_1.7.x

Happy to use it here too

Comment (n13) [reply scripts/compile.js:23]: @IanMendozaJaimes I am using local, so will add this in the local command too

@josejulio I had some issues with docker images in the past but that is 2 years ago - chances are docker is OK now on mac ;)

Comment (n13) [reply scripts/compile.js:23]: Local install is much more performant though

File: .gitmodules (hunk 1 above)
Comment (n13) [reply .gitmodules:3]: I will try and make cMake work

File: scripts/compile.js (hunk 5 above)
Comment (josejulio) [reply scripts/compile.js:23]: It is! is just that the tools are only supported on ubuntu and something else (can't recall right now). 
Maybe this could check if eosio-cpp is found in path (using `which` or something else) and execute it and fallback to the docker.

File: include/decay.hpp (hunk 4 above)
Comment (IanMendozaJaimes) [reply include/decay.hpp]: oh sorry for that, I was trying to see if that was causing my compilation error. I will change it again

File: .gitmodules (hunk 1 above)
Comment (n13) [reply .gitmodules:3]: cmake is now enabled and working

it's a lot slower than compile. but it works.

**Comment by n13** [comment]
Ok I adopted our compile scripts to work with cmake - should now work in both worlds the same. 

Cmake takes a lot longer to compile but I can live with that

I found that I don't need to use docker for running the EOSIO chain, I can keep running it locally, and that's great. I don't need docker for anything, actually. 

Speculative note to self: Possible that running cmake on docker sets up some cached locations of compile and other things. And then, when running it local on the mac, these locations and tools exist here too but they don't have the issues that the Xcode tools have - they work just fine.

**Comment by n13** [comment]
Will try this another way
